### PR TITLE
Implement Docker discovery

### PR DESCRIPTION
**Pull request by xperimental**
Implementation for #8173

### CODE REVIEW

File: discovery/dockerswarm/docker.go [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+	}
+
+	d.client, err = client.NewClientWithOpts(opts...)
+	if err != nil {
+		return nil, fmt.Errorf("error setting up docker swarm client: %w", err)
+	}
+
+	d.Discovery = refresh.NewDiscovery(
+		logger,
+		"dockerswarm",
```
Comment (roidelapluie) [inline discovery/dockerswarm/docker.go:163]:
```suggestion
		"docker",
```

File: docs/configuration/configuration.md [hunk 2]
```diff
@@ -645,6 +649,94 @@ filtering nodes (using `filters`).
 See [this example Prometheus configuration file](/documentation/examples/prometheus-dockerswarm.yml)
 for a detailed example of configuring Prometheus for Docker Swarm.
 
+### `<docker_sd_config>`
+
+Docker SD configurations allow retrieving scrape targets from [Docker Engine](https://docs.docker.com/engine/) hosts without a Swarm configuration.
```
Comment (roidelapluie) [inline docs/configuration/configuration.md:654]:
```suggestion
Docker SD configurations allow retrieving scrape targets from [Docker Engine](https://docs.docker.com/engine/) hosts.
```

File: docs/configuration/configuration.md [hunk 3]
```diff
@@ -645,6 +649,94 @@ filtering nodes (using `filters`).
...
+host: <string>
+
+# Optional proxy URL.
+[ proxy_url: <string> ]
+
+# TLS configuration.
+tls_config:
+  [ <tls_config> ]
+
+# The port to scrape metrics from, when `role` is nodes, and for discovered
```
Comment (roidelapluie) [inline docs/configuration/configuration.md:687]:
```suggestion
# The port to scrape metrics from, when `role` is nodes, and for discovered
```
This is not accurate.
Should be `when no ports are exposed for a container`

File: docs/configuration/configuration.md [hunk 4]
```diff
@@ -645,6 +649,94 @@ filtering nodes (using `filters`).
...
+# tasks and services that don't have published ports.
+[ port: <int> | default = 80 ]
+
+# Optional filters to limit the discovery process to a subset of available
+# resources.
+# The available filters are listed in the upstream documentation:
+# Services: https://docs.docker.com/engine/api/v1.40/#operation/ServiceList
```
Comment (roidelapluie) [inline docs/configuration/configuration.md:694]: This is not the correct documentation link.

File: docs/configuration/configuration.md [hunk 5]
```diff
@@ -645,6 +649,94 @@ filtering nodes (using `filters`).
 See [this example Prometheus configuration file](/documentation/examples/prometheus-dockerswarm.yml)
 for a detailed example of configuring Prometheus for Docker Swarm.
 
+### `<docker_sd_config>`
```
Comment (roidelapluie) [inline docs/configuration/configuration.md:652]: This should be before dockerswarm_sd_config

File: docs/configuration/configuration.md [hunk 6]
```diff
@@ -206,6 +206,10 @@ digitalocean_sd_configs:
 dockerswarm_sd_configs:
   [ - <dockerswarm_sd_config> ... ]
 
+# List of Docker service discovery configurations.
+docker_sd_configs:
```
Comment (roidelapluie) [inline docs/configuration/configuration.md:210]: This needs to be also later on, in the alertmanager_config section.

File: docs/configuration/configuration.md [hunk 7]
```diff
@@ -645,6 +649,94 @@ filtering nodes (using `filters`).
...
+# resources.
+# The available filters are listed in the upstream documentation:
+# Services: https://docs.docker.com/engine/api/v1.40/#operation/ServiceList
+# Tasks: https://docs.docker.com/engine/api/v1.40/#operation/TaskList
+# Nodes: https://docs.docker.com/engine/api/v1.40/#operation/NodeList
+[ filters:
+  [ - name: <string>
+      values: <string>, [...] ]
+
+# The time after which the droplets are refreshed.
```
Comment (roidelapluie) [inline docs/configuration/configuration.md:701]:
```suggestion
# The time after which the containers are refreshed.
```

File: docs/configuration/configuration.md [hunk 8]
```diff
@@ -645,6 +649,94 @@ filtering nodes (using `filters`).
...
+  # It is mutually exclusive with `credentials`.
+  [ credentials_file: <filename> ]
+
+# Configure whether HTTP requests follow HTTP 3xx redirects.
+[ follow_redirects: <bool> | default = true ]
+
+```
+
+The [relabeling phase](#relabel_config) is the preferred and more powerful
+way to filter tasks, services or nodes. For users with thousands of tasks it
```
Comment (roidelapluie) [inline docs/configuration/configuration.md:732]:
```suggestion
way to filter containers. For users with thousands of containers it
```

File: docs/configuration/configuration.md [hunk 9]
```diff
@@ -645,6 +649,94 @@ filtering nodes (using `filters`).
...
+can be more efficient to use the Docker API directly which has basic support for
+filtering nodes (using `filters`).
```
Comment (roidelapluie) [inline docs/configuration/configuration.md:734]:
```suggestion
filtering containers (using `filters`).
```

File: discovery/dockerswarm/docker.go [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+		for _, v := range f.Values {
+			d.filters.Add(f.Name, v)
+		}
+	}
+
+	// There are other protocols than HTTP supported by the Docker daemon, like
+	// unix, which are not supported by the HTTP client. Passing HTTP client
+	// options to the Docker client makes those non-HTTP requests fail.
+	if hostURL.Scheme == "http" || hostURL.Scheme == "https" {
+		rt, err := config.NewRoundTripperFromConfig(conf.HTTPClientConfig, "dockerswarm_sd", false, false)
```
Comment (roidelapluie) [inline discovery/dockerswarm/docker.go:140]:
```suggestion
		rt, err := config.NewRoundTripperFromConfig(conf.HTTPClientConfig, "docker_sd", false, false)
```

File: discovery/dockerswarm/docker.go [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+			client.WithScheme(hostURL.Scheme),
+			client.WithHTTPHeaders(map[string]string{
+				"User-Agent": userAgent,
+			}),
+		)
+	}
+
+	d.client, err = client.NewClientWithOpts(opts...)
+	if err != nil {
+		return nil, fmt.Errorf("error setting up docker swarm client: %w", err)
```
Comment (roidelapluie) [inline discovery/dockerswarm/docker.go:158]:
```suggestion
		return nil, fmt.Errorf("error setting up docker client: %w", err)
```

**Review by roidelapluie — COMMENTED** [review]
1000 thanks!

Here is my initial review, I did not test it live yet.

File: docs/configuration/configuration.md (hunk 3 above)
Comment (xperimental) [reply docs/configuration/configuration.md]: I noticed that I completely missed the `port` option in my implementation. The value is currently ignored. Should I remove the `port` configuration option or add it as a fallback for containers which have no exposed ports?

**Review by xperimental — COMMENTED** [review]
Sorry for the "swarms" still left in the code, should have grepped before creating the PR.

Two additional questions:
- I noticed that the `droplets` is also mentioned in the `dockerswarm_sd_configs` documentation. Should I fix it there as well?
- Am I correct in assuming that you'll squash&merge the PR and so should not squash the code-review commits myself?

**Comment by roidelapluie** [comment]
> I noticed that I completely missed the port option in my implementation. The value is currently ignored. Should I remove the port configuration option or add it as a fallback for containers which have no exposed ports?

Can you please add it?

> Sorry for the "swarms" still left in the code, should have grepped before creating the PR.

Zero issue with this, this is why we do code review :)

> Am I correct in assuming that you'll squash&merge the PR and so should not squash the code-review commits myself?

Yes

**Comment by roidelapluie** [comment]
> Am I correct in assuming that you'll squash&merge the PR and so should not squash the code-review commits myself?

Yes please :)

File: docs/configuration/configuration.md (hunk 6 above)
Comment (xperimental) [reply docs/configuration/configuration.md:206]: I've added it to the other section as well. It first looked like other SDs are missing from the alertmanager_config section as well, but it turns out that the list is just sorted a bit differently. Is this something that you want fixed (possibly in a separate PR?)

Comment (roidelapluie) [reply docs/configuration/configuration.md:206]: Yes let's have it in a split PR after this one is merged :)

**Comment by roidelapluie** [comment]
Can you please add tests in the config package?

**Comment by acamilleri** [comment]
Thanks for the implementation ! I'm glad to hear that the policy on integrations has changed <3

Code looks good for me, i don't have the time now to test it, but ASAP.

**Review by roidelapluie — APPROVED** [review]

**Comment by roidelapluie** [comment]
Thanks!

**Pull request merged**